pkg/controller/migmigration/hooks.go

```diff
@@ -0,0 +1,214 @@
+package migmigration
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/types"
+
+	migapi "github.com/konveyor/mig-controller/pkg/apis/migration/v1alpha1"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8sclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func (t *Task) runHooks() (bool, error) {
+	var hookRef *corev1.ObjectReference
+	var client k8sclient.Client
+	var err error
+
+	switch t.Phase {
+	case PreBackupHooks:
+		hookRef = t.PlanResources.MigPlan.Spec.Hooks.PreBackup
+	case PostBackupHooks:
+		hookRef = t.PlanResources.MigPlan.Spec.Hooks.PostBackup
+	case PreRestoreHooks:
+		hookRef = t.PlanResources.MigPlan.Spec.Hooks.PreRestore
+	case PostRestoreHooks:
+		hookRef = t.PlanResources.MigPlan.Spec.Hooks.PostRestore
+	default:
+		err := fmt.Errorf("Unknown Hook Phase %s", t.Phase)
+		log.Trace(err)
+		return false, err
+	}
+
+	hook := migapi.MigHook{}
+	job := &batchv1.Job{}
+
+	if hookRef != nil {
+		err = t.Client.Get(
+			context.TODO(),
+			types.NamespacedName{
+				Name:      hookRef.Name,
+				Namespace: hookRef.Namespace,
+			},
+			&hook)
+		if err != nil {
+			log.Trace(err)
+			return false, err
+		}
+
+		switch hook.Spec.TargetCluster {
+		case "destination":
+			client, err = t.getDestinationClient()
+			if err != nil {
+				log.Trace(err)
+				return false, err
+			}
+		case "source":
+			client, err = t.getSourceClient()
+			if err != nil {
+				log.Trace(err)
+				return false, err
+			}
+		default:
+			err := fmt.Errorf("targetCluster must be 'source' or 'destination'. %s unknown", hook.Spec.TargetCluster)
+			log.Trace(err)
+			return false, err
+		}
+
+		if hook.Spec.Playbook == "" {
+			job = t.baseJobTemplate(hook)
+		} else {
+
+			configMap, err := t.configMapTemplate(hook)
+			if err != nil {
+				log.Trace(err)
+				return false, err
+			}
+
+			phaseConfigMap, err := hook.GetPhaseConfigMap(client, t.Phase)
+			if phaseConfigMap == nil && err == nil {
+
+				err = client.Create(context.TODO(), configMap)
+				if err != nil {
+					log.Trace(err)
+					return false, err
+				}
+			} else if err != nil {
+				log.Trace(err)
+				return false, err
+			}
+			job = t.playbookJobTemplate(hook, configMap.Name)
+		}
+
+		runningJob, err := hook.GetPhaseJob(client, t.Phase)
+		if runningJob == nil && err == nil {
+			err = client.Create(context.TODO(), job)
+			if err != nil {
+				log.Trace(err)
+				return false, err
+			}
+			return false, nil
+		} else if err != nil {
+			log.Trace(err)
+			return false, err
+		} else if runningJob.Status.Failed >= 5 {
```

Comment:
Just a note here as to why 5. The backoffLimit for jobs is 5. It's stated that jobs can fail for a number of reasons, like OOM, being evicted from a node, etc. Jobs will continue to spawn and attempt to reach success until they reach the the limit.

The default value for spec.backoffLimit is 6, which as far as I can se actually seems to only allow up to 5 failures:

```
status:
  conditions:
  - lastProbeTime: "2020-04-01T20:48:01Z"
    lastTransitionTime: "2020-04-01T20:48:01Z"
    message: Job has reached the specified backoff limit
    reason: BackoffLimitExceeded
    status: "True"
    type: Failed
  failed: 5
  startTime: "2020-04-01T20:42:01Z"
```